### PR TITLE
Added YOLOv5 serverless function for auto labeling

### PR DESCRIPTION
**Pull request by wartek69**
### Motivation and context
Integration of YOLOv5 as a serverless nuclio function that can be used for auto-labeling. Ultralytics has released YOLOv5 a while ago and it's one of the most used computer vision libraries and therefore it would make sense to support it in CVAT. The integration is quite simple into CVAT as Ultralytics maintains a Pytorch Hub model (https://github.com/ultralytics/yolov5/issues/36) and a docker image (https://hub.docker.com/r/ultralytics/yolov5).
Related issues from ultralytics repo: https://github.com/ultralytics/yolov5/issues/5427

### How has this been tested?
Automatic annotation was run using YOLOv5 on a custom dataset.
The serverless function was deployed using 
```
nuctl deploy --project-name cvat \
  --path serverless/pytorch/ultralytics/yolov5/nuclio \
  --volume `pwd`/serverless/common:/opt/nuclio/common \
  --platform local
```
Then using the 'Automatic annotation' action the function was tested and the auto-generated labels were controlled to check that no coordinates misfit is happening.

### Checklist

- [ x] I submit my changes into the `develop` branch
- [ x] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ x] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ x] I have added tests to cover my changes
- [ x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by nmanovic** [comment]
@wartek69 , thanks for the contribution! Could you please add information about the model into README as well? https://github.com/openvinotoolkit/cvat#deep-learning-serverless-functions-for-automatic-labeling

File: serverless/pytorch/ultralytics/yolov5/nuclio/main.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import json
+import base64
+from PIL import Image
+import io
+import yaml
```
Comment (nmanovic) [inline serverless/pytorch/ultralytics/yolov5/nuclio/main.py:5]: The import is unused (pylint complains about it). Could you please remove it?

Comment (wartek69) [reply serverless/pytorch/ultralytics/yolov5/nuclio/main.py]: the yaml import was the unused one, I've removed it!

**Comment by wartek69** [comment]
@nmanovic I've added the model to the readme!

**Review by nmanovic — APPROVED** [review]
LGTM, thanks for the contribution! Many users asked about YOLOv5 support.

**Comment by simaiden** [comment]
I would like to add that if you want to deploy a custom model you should:

- Copy the weights.pt file to /serverless/common
- Edit main.py file with :

```python
#Line 11
model_path = "/opt/nuclio/common/weights.pt"
model = torch.hub.load('ultralytics/yolov5', 'custom', path = model_path)  # or yolov5m, yolov5l, yolov5x, custom
```
- It's important that the classes names defined in function.yaml file match with the classes defined when you train the model.

**Comment by martinerk0** [comment]
@simaiden Wow, thank you it works! From docs, it seemed I had to download weights into docker image or something, this was much nicer. @nmanovic It would be useful to put this into docs, usually people would want to run YOLO with custom weights without doing anything with docker/nuclio.

**Comment by nmanovic** [comment]
@martinerk0 , could you please send us a PR with necessary updates for documentation?

**Comment by livan3li** [comment]
> ### Motivation and context
> Integration of YOLOv5 as a serverless nuclio function that can be used for auto-labeling. Ultralytics has released YOLOv5 a while ago and it's one of the most used computer vision libraries and therefore it would make sense to support it in CVAT. The integration is quite simple into CVAT as Ultralytics maintains a Pytorch Hub model ([ultralytics/yolov5#36](https://github.com/ultralytics/yolov5/issues/36)) and a docker image (https://hub.docker.com/r/ultralytics/yolov5). Related issues from ultralytics repo: [ultralytics/yolov5#5427](https://github.com/ultralytics/yolov5/issues/5427)
> 
> ### How has this been tested?
> Automatic annotation was run using YOLOv5 on a custom dataset. The serverless function was deployed using
> 
> ```
> nuctl deploy --project-name cvat \
>   --path serverless/pytorch/ultralytics/yolov5/nuclio \
>   --volume `pwd`/serverless/common:/opt/nuclio/common \
>   --platform local
> ```
> 
> Then using the 'Automatic annotation' action the function was tested and the auto-generated labels were controlled to check that no coordinates misfit is happening.
> 
> ### Checklist
> * [ x] I submit my changes into the `develop` branch
> * [ x] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
> * [ x] I have updated the [documentation](https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
> * [ x] I have added tests to cover my changes
> * [ x] I have linked related issues ([read github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
> * [ x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
>   [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))
> 
> ### License
> * [ x] I submit _my code changes_ under the same [MIT License](https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
>   Feel free to contact the maintainers if that's a concern.
> * [x ] I have updated the license header for each file (see an example below)
> 
> ```python
> # Copyright (C) 2022 Intel Corporation
> #
> # SPDX-License-Identifier: MIT
> ```

hi

how did you rund `nuctl deploy --project-name cvat \
  --path serverless/pytorch/ultralytics/yolov5/nuclio \
  --volume `pwd`/serverless/common:/opt/nuclio/common \
  --platform local` this comman. 
  
  i cannot run it on windows 10. It throws an errors like this: 
  
  `C:\Users\livan\cvat>nuctl deploy --project-name cvat \

Error - exec: "/bin/sh": file does not exist
    /nuclio/pkg/cmdrunner/shellrunner.go:96

Call stack:
stdout:

stderr:

    /nuclio/pkg/cmdrunner/shellrunner.go:96
Failed to get docker version
    /nuclio/pkg/dockerclient/shell.go:729
No docker client found
    /nuclio/pkg/dockerclient/shell.go:83
Failed to create docker client
    .../pkg/containerimagebuilderpusher/docker.go:31
Failed to create containerimagebuilderpusher
    /nuclio/pkg/platform/local/platform.go:88`

**Comment by nfrvnikita** [comment]
> Я хотел бы добавить, что если вы хотите развернуть пользовательскую модель, вам следует:
> 
> * Скопируйте weights.pt файл в / бессерверный / общий
> * Редактировать main.py файл с :
> 
> ```python
> torch
> = model "/opt/nuclio/common/weights.pt"
> = model_path #Строка 11.hub.load('ultralytics/yolov5', 'пользовательский', путь  = model_path) # или yolov5m, yolov5l, yolov5x, пользовательский
> ```
> 
> * Важно, чтобы имена классов были определены в функции.файл yaml соответствует классам, определенным при обучении модели.

Hi, how i can copy my weights to /serversless/common?
docker cp?

**Comment by simaiden** [comment]
> > Я хотел бы добавить, что если вы хотите развернуть пользовательскую модель, вам следует:
> > 
> > * Скопируйте weights.pt файл в / бессерверный / общий
> > * Редактировать main.py файл с :
> > 
> > ```python
> > torch
> > = model "/opt/nuclio/common/weights.pt"
> > = model_path #Строка 11.hub.load('ultralytics/yolov5', 'пользовательский', путь  = model_path) # или yolov5m, yolov5l, yolov5x, пользовательский
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * Важно, чтобы имена классов были определены в функции.файл yaml соответствует классам, определенным при обучении модели.
> 
> Hi, how i can copy my weights to /serversless/common? docker cp?

Just copy in your os filesystem, cvat mount the entire folder into the docker container

**Comment by chang50961471** [comment]
How to clean up disk space by automatically annotating whether it will occupy a large amount of disk space

**Pull request merged**